src/errors/error_kind.rs

```diff
@@ -0,0 +1,319 @@
+    )]
+    AcceptEncodingUnnegotiable,
+    #[error("ETag header was invalid")]
+    ETagInvalid,
+    #[error("Age header was invalid: length out of bounds (unsized 64 bit integer)")]
+    AgeInvalid,
+    #[error("Clear-Site-Data header was invalid: {}", .0)]
+    ClearSiteDataInvalid(std::string::ParseError),
+    #[error("Cache-Control header was invalid")]
+    CacheControlInvalid,
+    #[error("Authorization header was invalid: {}", .0)]
+    AuthorizationInvalid(AuthError),
+    #[error("WWW-Authenticate header was invalid: {}", .0)]
+    WWWAuthenticateInvalid(&'static str),
+    #[error("Content-Location header url was invalid: {:?}", .0)]
+    ContentLocationInvalidUrl(url::ParseError),
+    #[error("Content-Location header base url was invalid: {:?}", .0)]
+    ContentLocationInvalidBaseUrl(Box<dyn std::fmt::Debug + Send + Sync + 'static>),
+    #[error("Expect header was malformed.")]
+    ExpectInvalid,
+    #[error("Timing-Allow-Origin header was invalid: {:?}", .0)]
+    StrictTransportSecurityInvalid(&'static str),
+}
+
+impl From<HeaderError> for Error {
+    fn from(other: HeaderError) -> Self {
+        Error::Header(other)
+    }
+}
+
+#[derive(Debug, Diagnostic, ThisError)]
+#[non_exhaustive]
+pub enum AuthError {
+    #[error("`{}` Auth had invalid credentials: {}", .0, .1)]
+    CredentialsInvalid(AuthenticationScheme, &'static str),
+    #[error("`{}` is not a recognized auth scheme.", .0)]
+    SchemeUnrecognized(String),
+    #[error("Could not find auth scheme")]
+    SchemeMissing,
+    #[error("Could not find auth credentials")]
+    CredentialsMissing,
+    #[error("Expected `{}` auth scheme but found `{}`", .0, .1)]
+    SchemeUnexpected(AuthenticationScheme, String),
+    #[error("Could not find www-auth realm")]
+    RealmMissing,
+}
+
+impl From<AuthError> for Error {
+    fn from(other: AuthError) -> Self {
+        Error::Header(other.into())
+    }
+}
+
+impl From<AuthError> for HeaderError {
+    fn from(other: AuthError) -> Self {
+        HeaderError::AuthorizationInvalid(other)
+    }
+}
+
+#[derive(Debug, Diagnostic, ThisError)]
+#[non_exhaustive]
+pub enum DateError {
+    #[error("HTTP Date-Time not in {} format", .0)]
+    FormatInvalid(&'static str),
+
+    #[error("HTTP Date-Time failed all parsings: imf_fixdate, rfc850, asctime")]
+    Unparseable,
+
+    #[error("HTTP Date-Time invalid: parts out of logical bounds")]
+    OutOfBounds,
+
+    #[error("HTTP Date-Time string was not ASCII")]
+    NotASCII,
+
+    // Individual parts
+    #[error("HTTP Date-Time invalid seconds")]
+    SecondsInvalid,
+    #[error("HTTP Date-Time invalid minutes")]
+    MinutesInvalid,
+    #[error("HTTP Date-Time invalid hours")]
+    HourInvalid,
+    #[error("HTTP Date-Time invalid day")]
+    DayInvalid,
+    #[error("HTTP Date-Time invalid month")]
+    MonthInvalid,
+    #[error("HTTP Date-Time invalid year")]
+    YearInvalid,
+    #[error("HTTP Date-Time invalid week-day")]
+    WeekdayInvalid,
```

Comment:
Is anyone going to use individual variants like this, rather than a more general parse error? When would someone want to treat these programmatically differently, rather than just showing a different error?

Reply:
Hmm. The existing code just made all these errors and I kinda replicated it... I don't know if it would be useful... if not these could probably be `&'static str`s...